src/docs/blog/thoughts_on_linting.md

```diff
@@ -0,0 +1,208 @@
+
+    - It doesn't generalize to non-autofixing tasks (linters)
+
+For these two reasons, it makes sense to use an output directory to communicate file-level information:
+
+- Suggested edits to the code (formatting suggestions)
+
+- Reported errors in the code (linting errors)
+
+Note that the behaviour of editing source files inplace may be desirable. It may also be desirable to expose it in pants. See [[Linters, adapters and views|pants('src/docs/blog:thoughts_on_linting')#linters-adapters-and-views]] 
+
+The linting tool may use a certain directory structure under the output directory to communicate which files the diff or error refer too.
+One possible scheme is as follows:
+```
+{output_dir}/{linter_name}/{relative_path}.{extension}
+```
+Where the relative_path would be the relative path to the file from the root of the repository and the extension could be `.error` to indicate an error report relating to that input file or `.fixed` to indicate a suggested modification to that input file.
+Note that the `{linter_name}` isn't something I've actually implemented in my work on Twitter source code, but my script only ever ran one linter at a time and I would expect for pants to possibly run multiple linters on a given file at the same time. Having parallel directory structures would avoid overwriting the results of one linter with the results of another.
+
+#### Why not use RPCs?
+
+RPCs may be the right solution for tooling in the long term; for instance, if the Language Server Protocol was to be extended to add awareness of formatters/linters, it would almost certainly be the best way to communicate between tools.
+In the absence of an existing standard protocol, though; it feels like inventing a new RPC protocol to interact between linters and other tools would be a case of over-engineering.
+
+### Communication format
+
+Three different kinds of things may be communicated from a linter/formatter to the world:
+
+- a suggested edit in a file,
+
+    - For this, in my work at Twitter, I chose to simply output a file containing the modified version of the code
+
+    - This was because arcanist understands that format, which is a detail. If we were willing to spend a bit of time on it, arcanist could easily be made to understand a `diff` by simply applying it to the original file and working with the resulting file
+
+    - If I had to recommend a way to do it, I would say that outputting a `diff` is more elegant
+
+- a reported linting error,
+
+    - To report a linting error to a tool, one of the easiest way is json. It's simple to output and to parse in most languages. It reads better and parses more easily than XML (*opinion)
+
+    - If the linter doesn't provide a rich enough output, falling back to plain text may also be OK; providing that the tools which parse this output are aware of this detail.
+
+    - Here is a suggested json schema that would contain most information ever needed:
```

Comment:
LSP _does_ define a schema for a "Diagnostic": https://microsoft.github.io/language-server-protocol/specification#diagnostic

It uses a range (start and end) rather than just a line+col, and uses "severity" rather than "level", but otherwise looks unsurprisingly similar :)

Reply:
Nice catch! My google-fu didn't yield this :)

Reply:
I mentioned it in #9034